### PR TITLE
RFD 26 - Custom Approval Conditions

### PR DESCRIPTION
**Pull request by fspmarshall**
[Rendered](https://github.com/gravitational/teleport/blob/fspmarshall/rfd-custom-approval-conditions/rfd/0026-custom-approval-conditions.md)

### CODE REVIEW

**Review by klizhentas — COMMENTED** [review]
I like the idea of `proposals` as a vote log and then delayed evaluation of proposals, because
client users can build their own evaluators of proposals.

I think we need more work on the proposal evaluator, how do we express the following:

* Alice can only approve requests for her team
* Bob needs at least two approvers from his team, but can nominate more as a reviewer
* Bob can approve but not override the roles proposed by Carol
* Request will suggest Alice and bob

Also, how do we query:

* Who can approve requests for this requests?

Building on your idea of access_request generators, we can define a predicate that gets evaluated on every state transition.

Default predicate (for backwards compatibility) will approve the request if there is any proposal moving it to the approved state.

We can define other predicates that take the list of proposals and define true or false - saying whether the request is approved. They can take the access request and say things liks:

```yaml
; default predicate approves if there is any proposal moving it to approved state
approve_if: any(req.proposals, 'approved')
```

At least two people have approved:

```
approve_if: min_approvers(2)
```


At least two reviewers from my team has approved my request:

```yaml
approve_if: ge(has_labels(req.proposals, user.traits.team), 2)
```

Nop request that lets clients implement their own logic and with it state machine will do nothing, never approving any request

```yaml
approve_if: nop()
```

This borrows the same idea from predicates, we'd need to define predicates that work out of the box.

**Review by a-palchikov — COMMENTED** [review]
so far, a couple of typos.

File: rfd/0014-custom-approval-conditions.md
Comment (a-palchikov) [inline rfd/0014-custom-approval-conditions.md:271]: If `dave`'s approval requests were to be replicated for a single role giving 3 requests for `foo`, `bar` and `bin` respectively with the corresponding threshold (`2`), it could be easier to reason about. In this case, `bob` and `alice` would both be able to approve `foo` and `bar` but `bin` but would be left unapproved, so `dave` would have to wait for someone else with a permission to approve `bin` to successfully gain access.

Comment (klizhentas) [reply rfd/0014-custom-approval-conditions.md:271]: This would make the processing much more complex though - right now the request is a transaction with a state, it's either approved or not. With your proposal, the request would be split into multiple object, each with it's own state. This would be quite hard to reason about, as request could be partially approved/denied.

**Comment by awly** [comment]
Can this be merged, since the implementation is merged already?

**Review by russjones — APPROVED** [review]

**Pull request merged**
